File: problem004.py
def is_palindrome(s):
    if len(s) <= 1:
        return True
    if s[0] != s[-1]:
        return False
    return is_palindrome(s[1:-1])

print sorted(x_1*x_2 for x_1 in xrange(999, 99, -1)
                     for x_2 in xrange(999, 99, -1)
                     if is_palindrome(str(x_1 * x_2)))[-1]

File: problem004.rb
def is_palindrome(s)
  if s.length <= 1
    return true
  end
  if s[0] != s[-1]
    return false
  end
  return is_palindrome(s[1..-2])
end

three_digits_pal = Array.new
999.downto 100 do |x_1|
  999.downto 100 do |x_2|
    if is_palindrome((x_1*x_2).to_s)
      three_digits_pal.push(x_1 * x_2)
    end
  end
end

puts three_digits_pal.sort()[-1]

